src/simple_controller_impl/src/main.rs

```diff
@@ -0,0 +1,204 @@
+async fn reconcile(cr: Arc<SimpleCR>, ctx: Arc<Data>) -> Result<Action, Error> {
+    let client = &ctx.client;
+
+    let cr_name = cr
+        .metadata
+        .name
+        .as_ref()
+        .ok_or_else(|| Error::MissingObjectKey(".metadata.name"))?;
+    let cr_ns = cr
+        .metadata
+        .namespace
+        .as_ref()
+        .ok_or_else(|| Error::MissingObjectKey(".metadata.namespace"))?;
+
+    // Prepare the input for calling reconcile_core
+    let cr_key = ResourceKey {
+        name: cr_name.clone(),
+        namespace: cr_ns.clone(),
+    };
+    let mut state = reconcile_init_state();
+    let mut resp_option: Option<APIResponse> = None;
+
+    // Call reconcile_core in a loop
+    loop {
+        // If reconcile core is done, then breaks the loop
+        if reconcile_done(&state) {
+            break;
+        }
+        // Feed the current reconcile state and get the new state and the pending request
+        let (state_prime, req_option) = reconcile_core(&cr_key, &resp_option, &state);
+        // Pattern match the request and send requests to the Kubernetes API via kube-rs methods
+        match req_option {
+            Some(req) => match req {
+                APIRequest::GetCRReq(key) => {
+                    let cr_api = Api::<SimpleCR>::namespaced(client.clone(), &key.namespace);
+                    resp_option = Some(APIResponse::GetCRResp(cr_api.get(&key.name).await));
+                }
+                APIRequest::CreateConfigMapReq(cm) => {
+                    let cm_api = Api::<ConfigMap>::namespaced(
+                        client.clone(),
+                        cm.metadata
+                            .namespace
+                            .as_ref()
+                            .ok_or_else(|| Error::MissingObjectKey(".metadata.namespace"))?,
+                    );
+                    let pp = PostParams::default();
+                    resp_option = Some(APIResponse::CreateConfigMapResp(
+                        cm_api.create(&pp, &cm).await,
+                    ));
+                }
+            },
+            _ => resp_option = None,
+        }
+        state = state_prime
+    }
+
+    Ok(Action::requeue(Duration::from_secs(100)))
+}
+
+// The controller triggers this on reconcile errors
+fn error_policy(_object: Arc<SimpleCR>, _error: &Error, _ctx: Arc<Data>) -> Action {
+    Action::requeue(Duration::from_secs(1))
+}
+
+// Data we want access to in error/reconcile calls
+struct Data {
+    client: Client,
+}
+
+// The main function registers the (1) the reconcile function and (2) the trigger of reconcile
+#[tokio::main]
+async fn main() -> Result<()> {
+    tracing_subscriber::fmt::init();
+    let client = Client::try_default().await?;
+
+    let crs = Api::<SimpleCR>::all(client.clone());
+
+    info!("starting simple-controller");
+
+    Controller::new(crs, ListParams::default()) // The controller's reconcile is triggered when a CR is created/updated
```

Comment:
The set up here with triggers determines when `reconcile_core()` is invoked. That is, this is a key assumption in our specs. Any thoughts on how to link it?

Reply:
It should be doable. The difficulty is to match the trigger condition written in a predicate to the imperative method sequences that set up the controller here.